Execute commands not yet been tested with Simulink

Harjaapb opened the pull request.

Commit:
Establish template of command library

Commit:
Establish template of command library

Commit:
Add command and validation template

Commit:
Add command and validation template

Commit:
Fix slow performance when stream of serial comes

Commit:
Fix slow performance when stream of serial comes

Commit:
Change digital write from toggle to actually write

Commit:
Change digital write from toggle to actually write

Commit:
Create new type for channel to split port and pin

Commit:
Create new type for channel to split port and pin

Commit:
Demo command with 2M baud :D

Commit:
Merge branch 'Command' of https://github.com/DAQSimple/DSX_fw into Command

Commit:
Incomplete command library

Commit:
Merge branch 'develop' into Command

Commit:
merge develop into Command

Harjaapb committed:
PWM execute functions

- Removed the validation checks in the PWM.c
- Added in the remaining digital input pins in channel.c
- Added PWM write and set frequency functions in validate.c and command.c
- Changed the int values for the loc to the board define macros

Harjaapb committed:
Added timer handles in command.c

- Removes the error

Harjaapb committed:
Merge branch 'develop' into exec_command

Harjaapb committed:
Commands updated with encoder library

- Added return types for Simulink verification of serial commands
- Changed loc integers to board define variables in channel.c and command.c
- Added encoder read
- Added the corresponding serial transmit to Simulink once the command has been executed
- Changed max duty cycle value to 99 because period is 100-1

Harjaapb committed:
Added Servo library + other minor changes

- Added the servo library which operates at 50 Hz and between duty cycle of 2.5%-12.5%
- Setting the Auto Reload Register (Period) inside both PWM and Servo libraries because servo uses a value of 1000-1 and PWM uses value of 100-1
- Conversion from degrees to duty cycle equation is DC = 25 + degrees/1.8
- Changed the Set_PWM_Level ID to Limit_switch (not sure if we'll be using either anyways)
- Commented out DI7 and DI8 (Can include later on if we want those pins to have alternate function to limit switches)
- Changed the validation for the PWM frequency to include the checks for user input for Hz and 10kHz
- Changed the validation for DAC to account for the negative and positive voltage
- Changed the PWM init function so that it initializes the timer instead, and the PWM starts in the updateDutyCycle function
- Changed max duty cycle to 99 because in reality it is 100-1

Harjaapb committed:
Update the PWM frequency command

- Uncommented DI7 and DI8 under channel.c
- Changed the command for the PWM set frequency. Max frequency is now 20 kHz so the val will contain first 4 significant digits and ret will have last significant digit if need be

Harjaapb committed:
Changing Sign of PWM Frequency

- Changed the sign from 1 & 2 -> 0 & 1

Harjaapb committed:
Add Servo library into command header

Harjaapb committed:
Minor changes

- Commented out safety_init function in main. Will be tested in another branch
- Commented out return pings which send a 9 back to Simulink

Harjaapb committed:
Comment out return ping

jaypacamarra commented on dsx-fw/DSX_Library/Command/Src/command.c:
```diff
@@ -0,0 +1,181 @@
+void execute_command(volatile DSX_data_t *dsx_data)
+{
+	// Validate dsx_data
+	if(!is_valid(dsx_data)) return;
```
If the dsx data is invalid then we should also set serial_available = false again and reset the dsx data and then return.

jaypacamarra commented on dsx-fw/DSX_Library/Command/Src/command.c:
```diff
@@ -0,0 +1,181 @@
+void cmd_encoder_read(volatile DSX_data_t *dsx_data){
+	dsx_data->sign = Encoder_Get_DIR();
+	dsx_data->val = Encoder_Read_RPM();
+	dsx_data->ret = RETURN_ENCODER_SPEED;
```
It doesn't make sense that the return type is RETURN_ENCODER_SPEED when the packet sent back to Simulink contains both speed and direction. This is me just being picky.

Harjaapb committed:
Change baud rate to match Simulink

jaypacamarra commented on dsx-fw/DSX_Library/Command/Src/validate.c:
```diff
@@ -0,0 +1,164 @@
+}
+
+// Validate Read Encoder Speed and Direction Command
+bool validate_read_encoder(volatile DSX_data_t *dsx_data){};
```
You forgot to add logic for validating a command to read the encoder

Harjaapb replied:
I added validation for the sign but do we need to check if the speed does not exceed some maximum value?

jaypacamarra replied:
I looked at the lookup table for encoder read. It looks like every parameter except ID is don't care so you don't need to check for anything. 

You don't need to check the speed or the sign because those are not required in the command to read the encoder. Simply returning true is okay in validate_read_encoder() based on the lookup table for encoder read.

Edit:
The loc parameter in the lookup is empty, is this don't care?

Harjaapb replied:
This is why the validation was left empty cause there was nothing to check for. There aren't options for encoder ports as the two pins are only dedicated to the encoder so I don't think we need to checks for the locs. Also there are no board defines for the encoder, at least in my branch.

jaypacamarra replied:
@Harjaapb validate_read_encoder needs to return true. or we'll never be able to read the encoder, that's what I'm saying.

jaypacamarra replied:
They way you have it, validate_read_encoder will always return false.

jaypacamarra commented:
When I try to do a PWM write in the firmware, it looks like its doing a servo write. Please fix this. See my comment below.

jaypacamarra commented on dsx-fw/DSX_Library/PWM/Src/PWM.c:
```diff
@@ -1,47 +1,35 @@
 
-// Function for updating the PWM duty cycle
+// Function for updating the PWM duty cycle and starting the PWM
 void updateDutyCycle (TIM_HandleTypeDef htim, uint8_t dutyCycle)
```
In your servo library, the writeServo function changes the ARR value to SERV_PERIOD. You need to change the ARR value  also in this updateDutyCycle function or else when you use writeServo and then updateDutyCycle , the output will not be right. I confirmed this on the scope.

jaypacamarra commented:
Also wanted to point out from testing what I obvserved:
PWM Write with 100, observed correct duty cycle of 100
Servo Write with 90, observed duty cycle of 7.2% with period of 19.92 ms
PWM Write with 100, observed incorrect duty cycle of 99

Harjaapb committed:
Further minor updates in validation and PWM

- changed RETURN_ENCODER_SPEED to RETURN_ENCODER_VELO
- Commented out CMD_EXECUTED return value to Simulink
- Resets dsx_data struct and set serial_available to false if the serial data coming in is invalid
- Added validation for read encoder
- Added the setting of the ARR register in the updateDutyCycle function to differetniate between servoWrite and PWMWrite

Harjaapb committed:
Removed encoder validation

Harjaapb committed:
Update encoder validation

Harjaapb committed:
Changed baud rate

Set to 230400 as a standard, instead of 256000

jaypacamarra reviewed (approved):
Good work
